### PR TITLE
blog: add Darwin spelunking post

### PR DESCRIPTION
**Pull request by mihaip**

### CODE REVIEW

File: data/blog/darwin-spelunking.mdx [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+title: 'Improving Tailscale via Apple’s open source'
+date: '2023-04-13'
+tags:
+  - 'oss'
+authors: ['mihai']
+images: ['/images/darwin-spelunking/social.png']
+summary: 'Spelunking through the Darwin kernel and userland helps Tailscale work better on macOS and iOS.'
+---
+
+# Improving Tailscale via Apple’s Open Source
```
Comment (thisisparker) [inline data/blog/darwin-spelunking.mdx:11]: we can cut this—title's automatically included at the top of the article

File: data/blog/darwin-spelunking.mdx [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
```
Comment (thisisparker) [inline data/blog/darwin-spelunking.mdx:12]: What do you think about boosting the lead here up a bit? I'm mostly thinking just inverting this paragraph. Maybe something like:

The ability to peek under the hood of the operating system can be a powerful tool for debugging. “Spelunking Apple's Open Source,” a recent blog post by Daniel Jalkut, struck a very familiar chord with me, reminding me of satisfying sessions poring through WebKit internals with my previous employer. I've been able to continue that pattern at Tailscale, now more focused on Darwin (the operating system at the core of macOS and iOS), its kernel, and its userland.

Just recently, Apple's open source came in handy when debugging before the Tailscale 1.36 release. We were investigating...

File: data/blog/darwin-spelunking.mdx [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+We were investigating two network-interface related bugs – one where Tailscale traffic was [looping on itself](https://github.com/tailscale/tailscale/issues/5156) (instead of being passed through to the actual physical network connection) and another where Tailscale was [sending traffic over the phone’s cellular interface](https://github.com/tailscale/tailscale/issues/6565) (even though Wi-Fi was available). The [fix](https://github.com/tailscale/tailscale/pull/6566) for both of these involved more correctly “binding” Tailscale’s network requests to the active network interface (and re-binding them when it changes).
```
Comment (thisisparker) [inline data/blog/darwin-spelunking.mdx:15]: What do you think about boosting the lead here up a bit? I'm mostly thinking just inverting this paragraph. Maybe something like:

The ability to peek under the hood of the operating system can be a powerful tool for debugging. “Spelunking Apple's Open Source,” a recent blog post by Daniel Jalkut, struck a very familiar chord with me, reminding me of satisfying sessions poring through WebKit internals with my previous employer. I've been able to continue that pattern at Tailscale, now more focused on Darwin (the operating system at the core of macOS and iOS), its kernel, and its userland.

Just recently, Apple's open source came in handy when debugging before the Tailscale 1.36 release. We were investigating...

File: data/blog/darwin-spelunking.mdx [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
...rd – it is the interface that is [associated with the default route](https://github.com/tailscale/tailscale/blob/9c2ad7086c0daef9c914ef1e59d0142b50f789f3/net/interfaces/interfaces_bsd.go#L38-L51). However, when Tailscale is configured to [use an exit node](https://tailscale.com/kb/1103/exit-nodes/), it becomes the default route, thus putting us back in danger of looping traffic. We want to know what the interface would be if Tailscale was not present (but without actually disabling it, since that is [disruptive](https://bugs.chromium.org/p/chromium/issues/detail?id=1076619)). I initially had an approach that [lived in our closed-source repo](https://github.com/tailscale/tailscale/blob/42c3052006141aab7d2573b2019c996641384014/net/interfaces/interfaces_bsd.go#L85-L93) and relied on private APIs (that I should probably [plead the 5th](https://en.wikipedia.org/wiki/Fifth_Amendment_to_the_United_States_Constitution#Self-incrimination) instead of discussing further). Worse yet, we started to get reports that users were [having](https://github.com/tailscale/tailscale/issues/7184) [issues](https://github.com/tailscale/tailscale/issues/7188), we were sometimes detecting the wrong interface.
```
Comment (thisisparker) [inline data/blog/darwin-spelunking.mdx:17]: Nit but we should use an — em dash here. Our observed convention is with a space around them, as you have here.

File: data/blog/darwin-spelunking.mdx (hunk 3 above)
Comment (mihaip) [reply data/blog/darwin-spelunking.mdx]: Agreed that it reads better, incorporated something along those lines.

**Pull request merged**